### PR TITLE
js-oop

### PR DESCRIPTION
**Pull request by RomanovAleksander**
### Object-Oriented JavaScript

[Demo](https://romanovaleksander.github.io/frogger-game/) |
[Code base](https://github.com/RomanovAleksander/frogger-game)

The code is submitted in a dedicated feature branch.
Please, review.

Please, could advise how to fix the bug when the character disappears over enemy? I tried different coordinates, but nothing works correctly.

### CODE REVIEW

File: submissions/RomanovAleksander/js_oop/app.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    'images/char-horn-girl.png',
+    'images/char-pink-girl.png',
+    'images/char-princess-girl.png'
+];
+
+let Enemy = function (x, y) {
+    this.sprite = 'images/enemy-bug.png';
+    this.x = x;
+    this.y = y;
+    this.speed = Math.floor(Math.random() * 250) + 70;
```
Comment (OleksiyRudenko) [inline submissions/RomanovAleksander/js_oop/app.js]: `70` is also a magic value worth being a well named constant.
Check for other magic values across your code.

File: submissions/RomanovAleksander/js_oop/app.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+};
+
+Enemy.prototype.update = function (dt) {
+    if (this.x < convasWidh) {
+        this.x += this.speed * dt;
+    } else {
+        this.x = -blockWidth;
+    }
+
+    if (!(this.y + blockHeight < player.y ||
```
Comment (OleksiyRudenko) [inline submissions/RomanovAleksander/js_oop/app.js]: This must be a block responsible for collision detection.
It is important part of the app's business logic therefore worth being a dedicated method.

Re bug you mentioned in your initial post:
- characters image dimensions are quite different from visible part dims, need to explore and adjust your math
- the logic within `if` condition is confusing there must be `&&` somewhere

File: submissions/RomanovAleksander/js_oop/app.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        this.y > player.y + blockHeight ||
+        this.x + blockWidth < player.x ||
+        this.x > player.x + blockWidth)) {
+        player.x = initialX;
+        player.y = initialY;
```
Comment (OleksiyRudenko) [inline submissions/RomanovAleksander/js_oop/app.js]: `player` is a global variable. In OOP classes should be agnostic of any particular instances. Also globals should not be referred to within classes. The only exclusions are constants and library/engine/framework functions.

File: submissions/RomanovAleksander/js_oop/app.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
+const convasWidh = 505,
```
Comment (OleksiyRudenko) [inline submissions/RomanovAleksander/js_oop/app.js]: A typo `convas` => `canvas`

File: submissions/RomanovAleksander/js_oop/app.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+let imagesOfHeroes = [
+    'images/char-boy.png',
+    'images/char-cat-girl.png',
+    'images/char-horn-girl.png',
+    'images/char-pink-girl.png',
+    'images/char-princess-girl.png'
+];
+
+let Enemy = function (x, y) {
+    this.sprite = 'images/enemy-bug.png';
```
Comment (OleksiyRudenko) [inline submissions/RomanovAleksander/js_oop/app.js]: This string also worth defining as a constant.

**Comment by RomanovAleksander** [comment]
I finished.

File: submissions/RomanovAleksander/js_oop/app.js (hunk 5 above)
Comment (OleksiyRudenko) [reply submissions/RomanovAleksander/js_oop/app.js]: If you do not implement a recommendation, please, give a reason.

File: submissions/RomanovAleksander/js_oop/app.js [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    if (this.x < canvasWidh) {
+        this.x += this.speed * dt;
+    } else {
+        this.x = -blockWidth;
+    }
+
+    this.collisionDetection();
+};
+
+Enemy.prototype.collisionDetection = function () {
```
Comment (OleksiyRudenko) [inline submissions/RomanovAleksander/js_oop/app.js]: By convention function names should start with a verb, imperative mood.
Functions that return boolean normally start with `is`, 'has` or similar and can be read as questions.
Please, refactor.

File: submissions/RomanovAleksander/js_oop/app.js [hunk 7]
```diff
@@ -0,0 +1,105 @@
+const canvasWidh = 505,
```
Comment (OleksiyRudenko) [inline submissions/RomanovAleksander/js_oop/app.js]: A typo `Widh` => `Width`

**Comment by RomanovAleksander** [comment]
Done

File: submissions/RomanovAleksander/js_oop/app.js (hunk 5 above)
Comment (OleksiyRudenko) [reply submissions/RomanovAleksander/js_oop/app.js]: @RomanovAleksander It is still a string per se.
Just imagine that at some point as the game grows this file is split into multiple files for better maintainability.
One of files will contain all constants and if someone decides to change a sprite for enemy that someone will be wondering why the sprite image path is set not somewhere around other constants.

Comment (RomanovAleksander) [reply submissions/RomanovAleksander/js_oop/app.js]: I understand, I'm sorry, I'll fix it right now in my repository.

**Pull request merged**